includes/TingSearchCqlDoctor.class.inc

```diff
@@ -0,0 +1,248 @@
+    // handle reserved characters.
+    $this->escape_reserved_characters();
+    // Remove multiple whitespaces.
+    $this->cql_string = preg_replace('/\s+/', ' ', $this->cql_string);
+  }
+
+  /**
+   * Method to convert a string to strict cql.
+   *
+   * Basically this method adds quotes when needed.
+   *
+   * @param string $string
+   *   The search query.
+   * @return string
+   *   Cql compatible string.
+   */
+  public function string_to_cql() {
+    // hendle parantheses
+    $this->fix_paranthesis();
+    // handle qoutes
+    $this->fix_qoutes();
+    // format the string
+    $return = $this->format_cql_string();
+    return $return;
+  }
+
+  /**
+   * Format cql string.
+   *
+   * Use private members $pattern and $replace to replace not valid cql phrases with valid.
+   *
+   * @return string
+   *
+   */
+  private function format_cql_string() {
+    // Last check. All parts of cql string must be valid.
+    $valid = TRUE;
+    $parts = preg_split($this->get_cql_operators_regexp(), $this->cql_string);
+    foreach ($parts as $part) {
+      if (!$this->string_is_cql($part)) {
+        $valid = FALSE;
+        break;
+      }
+    }
+
+    // Explode string by whitespace.
+    $expressions = explode(' ', $this->cql_string);
+    // replace keys with phrases
+    if (!empty($this->pattern)) {
+      $expressions = preg_replace($this->pattern, $this->replace, $expressions);
+    }
+
+    if ($valid) {
+      // Implode by blank.
+      return implode(' ', $expressions);
+    }
+
+    //String is not valid; implode by and.
+    return implode(' and ', $expressions);
+  }
+
+  /**
+   * Enqoute forward slashes and '-'.
+   *
+   * @return nothing
+   *   Alters private member cql_string.
+   */
+  private function escape_reserved_characters() {
+    $this->cql_string = str_replace('/', ' "/" ', $this->cql_string);
+    $this->cql_string = str_replace('-', ' "-" ', $this->cql_string);
+  }
+
+  /**
+   * Get a key for replacement in string.
+   *
+   * @return string
+   */
+  private function get_replace_key() {
+    $key_prefix = 'zxcv';
+    return $key_prefix . self::$replace_key++;
+  }
+
+  /**
+   * Handle parantheses.
+   *
+   * Look lace parantheses in string. If any found and content is not
+   * strict cql; enqoute the lot.
+   *
+   * @return nothing.
+   *   Alters private member cql_string.
+   *
+   */
+  private function fix_paranthesis() {
+    $valid_cql = TRUE;
+    //Grab content in paranthesis.
+    preg_match_all('$\(([^\(\)]*)\)$', $this->cql_string, $phrases);
+
+    if (empty($phrases[1])) {
+      // No matching paranthesis.
+      return;
+    }
+    foreach ($phrases[1] as $phrase) {
+      if (!$this->string_is_cql($phrase)) {
+        $valid_cql = FALSE;
+        break;
+      }
+    }
+    if ($valid_cql) {
+      foreach ($phrases[0] as $phrase) {
+        $this->set_replace_pattern($phrase);
+      }
+    }
+    else {
+      // one or more of the parenthesis holds invalid cql; quote the lot.
```

Comment:
I don't known much about CQL, but should it not only be the invalid parenthesis placement that should be fixes and not all ?